tests/pubsub/testgossipsub.nim

```diff
@@ -33,13 +33,15 @@ proc waitSub(sender, receiver: auto; key: string) {.async, gcsafe.} =
     return
   # turn things deterministic
   # this is for testing purposes only
-  # peers can be inside `mesh`, not just `gossipsub`
+  # peers can be inside `mesh` and `fanout`, not just `gossipsub`
```

Comment:
Not as far as I remember, `fanout` and `mesh` are a subset of the `gossipsub` peers, so you might miss it from one of the two, I believe the safest is to check for those in `gossipsub` itself.

Reply:
BTW, this is a good source - https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md

Reply:
Well then implementation is wrong,
here we are popping them from that container https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub.nim#L270-L275

Reply:
and that is the reason I added those checks

Reply:
> BTW, this is a good source - https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md

Yeah I used that as reference, was a good read

Reply:
So the point I'm making here, is that you are checking for peers in each of the tables, which is not guaranteed to happen. It happens now, only because the tests were it's being used, both publishes and subscribe to all topics, but things change once we only publish to a topic, but don't subscribe to it. I'm asking for caution with using this sort of logic indiscriminately and we should probably name the method accordingly - like `waitForAllSubscribed` or something....

> Well then implementation is wrong,

What is wrong, can you be more specific?

Reply:
> What is wrong, can you be more specific?
They are being removed from `gossipsub` table when they are added to `mesh` or `fanout`

Also I get your point but because of the reason above tests would fail if I didn't hack it this way, I agree it's a hack tho, we should probably have a `waitSub` in pubsub it self not in the test.

Reply:
> we should probably have a waitSub in pubsub it self not in the test.
Not sure about this.

AFAIK, the moving of peers around si by design, but this might be a good moment to have someone else verify the implementation against the spec - note that when in doubt I always resorted to the reference Go implementation, so if there are discrepancies it'd be a good to also compare against the Go.

Reply:
Oh yeah don't get me wrong, it might be ok.
Just to make sure the test runs properly I definitely have to check all the collections, that's all.